databuilder/query_model.py

```diff
@@ -365,38 +359,83 @@ class DomainMismatchError(Exception):
     ...
 
 
-PATIENT_DOMAIN = object()
+def validate_input_domains(node):
+    # The domain of a Frame or Series can be thought of as the set of its primary keys.
+    # This determines which other Frames or Series it can be validly composed with. The
+    # rules we want to enforce are:
+    #
+    #  * One-row-per-patient data can be composed with anything: by design every table
+    #    has a patient_id column and so data defined purely in terms of patient_id
+    #    can be joined with anything.
+    #
+    #  * Many-rows-per-patient data drawn from different tables can never be combined
+    #    — at least, not until we add an explicit JOIN operation.
+    #
+    #  * Many-rows-per-patient data drawn from a single underlying table can be combined
+    #    as long as any filters applied to the data can be arranged in a linear
+    #    sequence. That is, it's legal to combine `Table A -> Filter B -> Filter C` and
+    #    `Table A -> Filter B`. But it's not legal to combine `Table A -> Filter D` and
+    #    `Table A -> Filter E`. There's never a need to use such constructions and they
+    #    will almost always be the result of user error. It's better to reject them
+    #    immediately than give the user unexpected results at query time.
+    #
+    # We enforce this by modelling domains as a hierarchy. At the root is the patient
+    # domain, each table has an unique domain descending from this, and each filter
+    # operation creates a new domain descending from the domain of its source. A valid
+    # operation is one where the domains of its inputs are "nested" in the sense that
+    # each domain is an ancestor or descendant of every other.
+    #
+    # We use sets to implement this hierarchy: we start with the patient domain as a set
+    # containing a single unique value, and we create descendent domains by forming the
+    # union of the parent domain with a new unqiue value. The `issubset()` relation then
+    # gives us the "is domain ancestor" semantics we want.
+    #
+    # Another way of expressing the "nested" condition above is that the input domains
+    # must be totally ordered by ancestory. We can check this by sorting them (sets
+    # naturally sort by subset-hood in Python) and then confirming that the ordering is
+    # total by testing that each adjacent pair has the required subset relation.
+    domains = sorted(get_input_domains(node))
+    for a, b in zip(domains, domains[1:]):
+        if not a.issubset(b):
+            raise DomainMismatchError(
+                f"Attempt to combine unrelated domains:\n{a} and {b}"
+                f"\nIn node:\n{node}"
+            )
```

Comment:
Maybe slightly cleaner if the detail of this message were part of the exception and we just pass in the data?

Reply:
Maybe. I think we are going to want to raise something more structured here (and the same for the `TypeErrors`) so that ehrQL can catch and translate these for its users. But I'd rather leave thinking about this until we do the work in ehrQL and can see what structure would be most helpful.

Reply:
Fine by me.